### DISCUSSION TITLE
Getting stanify feedback from stats audience

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#76] [people; relating]
## Preparing for Stan discourse post and/or Andrew's blog about stanify

@jandraor and @hyunjimoon will develop three casestudies with @tomfid, @hazhirr as well as Stan community's leaders. It aims for Bayes workflow on prey-predator, SEIR, WIP-inventory management model. 
 
# Role models:
- Aki's birthday problem on gaussian process workflow [here](https://avehtari.github.io/casestudies/Birthdays/birthdays.html) which builds eight different models depending on additional structure (long, short, event term) and distribution for variance parameter.

- Martin's [Simulation-Based Calibration Checking for Bayesian Computation:The Choice of Test Quantities Shapes Sensitivity](https://arxiv.org/pdf/2211.02383.pdf)

- Michael's https://betanalpha.github.io/assets/case_studies/gaussian_processes.html#43_Approximating_Gaussian_Processes
> The precision matrix can also be directly related to alternative representations of Gaussian processes, such as spectral densities and stochastic differential equations. Although these are more mathematically sophisticated representations they can offer a wealth of understanding about the functional behavior encoded within a Gaussian process, complementing the interpretability of the covariance function representation.

- Hazhir's [Heterogeneity and Network Structure in the Dynamics of Diffusion (Agent based vs compartment model)](https://pubsonline.informs.org/doi/abs/10.1287/mnsc.1070.0787)
> Comparing disaggregated DE models to AB models may be useful in designing stochastic compartment models that capture heterogeneity and network effects using the fewest additional compartments, if the network structure is known and stable. 3 Testing this proposal is beyond the scope of this paper.

- Angie's [Predicting Engine Failure With Hierarchical Gaussian Process](https://mc-stan.org/users/documentation/case-studies/gaussian-process.html) shows re-parameterization based on Stan's divergence plot

- King's [Avoidable errors in the modelling of outbreaks of emerging pathogens, with special reference to Ebola](https://royalsocietypublishing.org/doi/10.1098/rspb.2015.0347), model excluding demographic noise biases estimation; CTMC and SDE

# Format:
- for **term consistency** we will use the column of stan term defined in [stanify](https://github.com/Data4DM/stanify) which I plan to ask feedback from at least two community leaders from each side
- final format is [quarto](https://quarto.org/)
 
# Questions
1. effect of measurement noise and process noise
a. relative scale between the two noise and their priors 
b. meaning and formulation of process noise in hierarchical context

2. examples that HMC can pass (fit fast) and/or **fail with diagnostics (fail fast)** i.e. joker combination #71  degeneracy from colinearity, hierarchy, multimodality and its remedy (Michael's [managing degeneracy](https://betanalpha.github.io/assets/case_studies/identifiability.html#3_Managing_Degeneracies)) 

3.  [bad parameterization example detected by SBC](https://hyunjimoon.github.io/SBC/articles/bad_parametrization.html): mixed order in gaussian mixture, standard-deviation vs precision formulation

4. types of hierarchy #66 demarcating the role of Vensim and Stan

5. hierarchical compartment model 
- Niko's [monster-model revisited](https://github.com/nsiccha/monster#the-monster-model-revisited-using-stan)
- addressing computational challenges with adaptive precision and time step 

# Goal
1. Leverage the authority of two community leaders during preparation (Hazhir and Tom, Andrew and Aki or Michael) to communicate the importance of the need for time and structure convergence of estimator / generator to 1. [This](https://discourse.mc-stan.org/t/the-monster-model-hierarchical-ode-models/22683/16) stan post on hierarchical ode.

> In Stan community, $\frac{t_{est}}{t_{gen}} > 1$, but in dynamic simulation community, $\frac{t_{est}}{t_{gen}} << 1$ where  $t_{gen}, t_{est}$ are time spent in designing generator (parameter scaler values to state time-series matrix) and estimator (reverse of generator). From trustable model perspective, $\frac{t_{est}}{t_{gen}} \sim 1$ and symmetric generator and estimator are ideal. However, I discovered some model components that were asymmetrically focused in two communities. Topic of adaptable `time_step`, `identifiablity`, `diagnostics` are much valued in estimator-focused community where as `process noise`, `aggregation` get greater attention in generator-focused community. The difficulty of SBC lies in matching these asymmetry which were not surfaced when only one side of the joint model (joint of generator and estimator which forms joint distribution of parameter and observation) was the focus. 

2. extend the boundary of model (add computation to managerial-statistics for SD people, add managerial to statistics-computation for Stan people) e.g. different perception of time-step [asked in stan discourse](https://discourse.mc-stan.org/t/the-monster-model-hierarchical-ode-models/22683/14)

3. reparameterization strategies based on visual diagnostices for both decision-maker and computer
- e.g. from hierarchical gaussian process [casestudy](https://mc-stan.org/users/documentation/case-studies/gaussian-process.html): ![image](https://user-images.githubusercontent.com/30194633/200345910-8a10fa74-ac2b-42fc-a6c9-cb8e8fada300.png)

# Three case studies
## 1. prey-predator
a. the importance of tight prior: sensitive parameter retrieval
-- replicating degeneracy observed in [stanify](https://github.com/Data4DM/stanify/blob/main/stan_files/HW7_estag_mnoisebeta22_pnoisedot01/HW7_estag_mnoisebeta22_pnoisedot01_draws2data.stan) (two birth frac. estimation, m_noise as beta(2,2), pnoise scale .01 in generator) in vensim

b. importance of tight measurement distribution (negative-binomial (mean-variance linear VS lognormal (too fat-tail; power of 2; mean is exp( µ + 2 1 σ 2), the variance is exp(2 µ ) exp(σ 2 )(exp(σ 2 ) − 1))

c. effect of process noise (process noise have oscillation-killing, aggregation effect which leads to stable but too much aggregated estimation which can be thought as layer 1 (question of new subscript can be discussed; new ship with its engine type included in train set VS  new ship with its engine type not included in train set from the [paper](https://onlinelibrary.wiley.com/doi/full/10.1002/ail2.22) on korea navy engine failure)
<img width="493" alt="image" src="https://user-images.githubusercontent.com/30194633/200133747-055573e2-f9f5-4f68-8e91-4eff04888a0b.png">
d. clustered subgroup with normal mixture prior for population distribution

(Jair's comment: SDA is based on this model so changing dominant loop can be interesting)

## 2. SEIR
a. degeneracy from collinearity (mean generation time (1/gamma + 1/sigma), 2. basis reproduction number (gamma/beta))
b. interacting partial pooling (e.g. age group), ode function is called n_obs times
c. sampling distribution: negative binomial (poisson (overconfidence and biased estimates))

## 3. Inventory management 
a. degeneracy from separating compartments degeneracy from multimodality (Hazhir discovered a bad parameter recovery with hierarchical Bayesian with powell)

b. non-interacting partial pooling documented in https://github.com/Data4DM/BayesSD/issues/17, ode function is called n_obs * n_region times

c. effect of timestep (decrease the time step to make sure it is not integration error, exponential decay going to 0; stock for expected customer order)

## 4. Future connection 
- #86 modeling capability of SDE in Vensim or Stan? Or do we have a walkaround (Jair says CTMC)? Michael's [inferring SDE](https://betanalpha.github.io/assets/case_studies/stochastic_differential_equations.html#23_Inferring_Stochastic_Processes). [This](https://mc-stan.org/users/documentation/case-studies/gaussian-process.html) casestudy on hierarchical gaussian process.

- Ed's [Combining stock-and-ﬂow, agent-based, and social network methods to model team performance](https://onlinelibrary.wiley.com/doi/abs/10.1002/sdr.1613)
> the differences among the productivities of the diversity scenarios is not due to the number of members from each subgroup, which is constant among the three scenarios. Rather it is from members’ positions in the communications hierarchy.

**Comment by hyunjimoon** [comment]
How much quantitative data is needed to reliably infer parameters of an SD model?
to define three steps as follows:

1. parameterize SD model 
input: pre-processed differential equations (.mdl) • reparameterize to management friendly form: Tom's [parameter taxonomy](https://metasd.com/2022/09/parameter-taxonomy/) • reparameterize to statistical computation friendly form: express parameters using notations defined in [Bayesian taxonomy](https://github.com/Data4DM/BayesSD/discussions/5) 
output: structure (processed differential equations)

2. collect and construct driving data array
input: structure + pre-processed data (data lake)

    conditional on structure 
 • collect required data 
 • frame data as data array (labeled multidimensional array) 
 • frame prior knowledge on parameters as driving prior distribution 
output: consistent (structure, numeric)
3. flexible tests using experimental priors
input: (structure, numeric) + managerial knowledge   • choose test quantities: (log likelihood + log prior penalty) is default and modeler can add other test quantities • choose setting: different experimental priors for estimated, policy parameters
output: numeric and graphical diagnostics 
——
Iterate 1,2,3 and inspect diagnostics 
 • numeric: d(target_obs, target_simulated_obs), inferential and predictive empirical continuous rank probability score (defined in Michael’s [writing](https://betanalpha.github.io/assets/case_studies/modeling_sparsity.html#36_Comparison)), Rhat, effective sample size, p-loo, expected log predictive density for a new data point, E-BFMI • graphical: scattered divergence on pairs plot, sbc rank plotfinal output: useable consistent (structure, numeric, setting) 
——stanify draws2data2draws automate the third step:
```
def draws2data2draws(structure, numeric, setting,  S, M, N):
  model = vensim2stan(structure)  model.set_numeric(numeric)

  model.set_setting(setting, N)
  prior_sample = sample(experimental prior, S)
  target_simulated_obs = draws2data(model, prior_sample)

  for s in range(S):
    posterior_sample[s] = data2draws(model, target_simulated_obs[s], M)
   
  def draws2data(model, prior_sample)
    return generate(model, prior_sample).target_simulated_obs
  def data2draws(model, target_simulated_obs, M)
    return estimate(model, target_simulated_obs, M)

  def diagnose(prior_sample, posterior_sample, target_simulated, target_simulated_obs, test_quantity):
    return compare(test_quantity(prior_sample, target_simulated, target_simulated_obs, target_obs),
                              test_quantity(posterior_sample, target_simulated, target_simulated_obs, target_obs))
  
  return diagnose(prior_sample, posterior_sample, target_simulated, target_simulated_obs, target_obs, ('loglik'))
```
——
I wish to get feedback on the overall procedure defined above and two new design decisions:
 1. I removed the concept of assumed parameter as it can be framed as estimated with point mass prior
 2. I separated driving prior (step 2) from experimental prior (step 3). From my experience, Stan users spend most of their time tweaking priors until they get good enough diagnostics. This computational consideration should be separated from our real knowledge of a parameter.

**Reply by hyunjimoon** [reply]
<img width="958" alt="image" src="https://user-images.githubusercontent.com/30194633/199723355-a33c86b8-3ad9-43e9-ba0b-72ac292c2fed.png">

**Reply by hyunjimoon** [reply]
- Jair used reporting rate \rho \sim beta(2,2), \beta lognormal
- Angie used  beta(2,2) for measurement noise prior which caused divergence; perhaps it was too big considering normal(.1, .01) didn't cause divergence.

**Comment by jandraor** [comment]
## questions to ask to statistics experts via blog

What is the advantage of adding process noise to the SD prey-predator in comparison to the time-continuous Markov chain version?

**Comment by hyunjimoon** [comment]
Correlation endgoal
<img width="951" alt="image" src="https://user-images.githubusercontent.com/30194633/200335927-e13b9673-a4fa-42a2-9c55-3cd62a77f99e.png">

**Comment by hyunjimoon** [comment]
Both Stella and Vensim has output for xml. 
i) python -> `inferencedata` (netcdf) -> R -> `SBC`
ii) vensim -> netcdf -> R ->  `SBC`

Mike has a code that translates netcdf to rvars [here](https://github.com/stan-dev/posterior/issues/85#issuecomment-869939038). Explanation for [rvars](https://mc-stan.org/posterior/articles/rvar.html) by Matthew and SBC seems to support rvars from [here](https://github.com/hyunjimoon/SBC/issues/35#issuecomment-933001279).

**Comment by hyunjimoon** [comment]
Thank you, Andrew Gelman and Tom Fiddaman @tomfid for supporting this project. Below is what @Dashadower and I planned together. 

# Hierarchical ODE SBC

## Time range
12 weeks from Dec.W1 ~ FebW4.

## Goal 
1. Speed up
measure of success: 
a. parameters combination (which pair and how many) for an [inventory management model](https://github.com/Data4DM/stanify/blob/main/vignette/stan_files/Inven_S1N200Q2R1/functions.stan#L2470) whose estimation ends in .5 hour and pass arviz diagnostics (rank, sbc, rhat, khat)
b. at least one stan discourse question or asking feedback from ODE experts (Juho, Sebastian, Charles, Yi)

- trace bottleneck and experiment with what causes 13hours for inventory model estimation
- potential causes are less adaptive warmup ([remedy](https://discourse.mc-stan.org/t/the-monster-model-hierarchical-ode-models/22683/18) for the monster model by Niko), diffuse prior, hierarchy,  not using `reduce_sum` etc. Based on the auto-hierarchy stanfile translation and experiment setting Angie prepared, Shinyoung will experiment with different model settings to find the cause of the bottleneck and apply existing methodologies. 
- .5 hours is based on the fitting time from Niko's report
- https://github.com/Data4DM/BayesSD/discussions/89#discussioncomment-4264309

2. Connect
measure of success: 
a. arviz function supporting sbc plot of a hierarchical model
b. sbc ecdf_diff and coverage plot of applying a. to inventory model 
- the decision to connect to R vs doing all in python is deferred to the end of December; [this](https://github.com/hyunjimoon/SBC/issues/41#issuecomment-1312512398) contains blueprint for connection to R
- regardless of the decision, Angie will develop `inferencedata`-based sbc plot for a hierarchical model by the end of Dec. ~ mid Jan. with @OriolAbril
- Shinyoung will apply this plot (or the logic - if we decide to connect with `SBC` package) to the inventory model after 1's completion.

3. Action
measure of success
a. replicate [blog's output](https://twiecki.io/blog/2019/01/14/supply_chain/) on the saved cost by 29 ("Expected value of including uncertainty = 29.31")
b. add sbc, plot_ppc, pair plot and other basic diagnostics to blog's storyline(from 2)

- Angie will prepare vensim model by the end of Jan. Shinyoung and Angie will use this model to replicate blog after 1 and 2's completion.
- https://github.com/Data4DM/BayesSD/discussions/105

## Output
findings will be presented in quarto or notebook format which will be posted in Andrew's blog for feedback (as planned above).

**Reply by hyunjimoon** [reply]
Tom's suggestion on speedup: https://github.com/Data4DM/BayesSD/discussions/89#discussioncomment-4316551

**Reply by hyunjimoon** [reply]
If speed issue turns out to be relatively easy, plotting is another low-hanging fruit: https://github.com/nsiccha/monster/blob/a362738efd9c83525a2edb6db3e3646fba1a293f/plotting.py

**Reply by hyunjimoon** [reply]
@Dashadower and Angie discussed:
For 1: 
1. https://github.com/Data4DM/BayesSD/discussions/121
- with S:14, 19, plot `observed data` as Angie thinks (desired state - state)/adj_time being infinite is the problem
- sbc.observed_data.sel(`prior_draws` = 14) can be used
- notebook file: https://github.com/Data4DM/stanify/blob/main/vignette/inven_hier_wopnoise_flowobs.ipynb

2. N=200, improving if-else structure
i) N = 20 -> 300 ([nesting bracket statements error](https://github.com/Data4DM/BayesSD/discussions/124#discussioncomment-4363806))
ii) non hier -> hierarchical
iii) parameter from 2->8 (degeneracy will increase) 

3. problem cause will be degeneracy. how to measure speed improvements from scaling is introduced in
https://betanalpha.github.io/assets/case_studies/identifiability.html#522_Comparing_The_Performance_of_Scaling_Strategies

**Reply by hyunjimoon** [reply]
12W2
- loglik
- generating makefile for experiments https://opensource.com/article/18/8/what-how-makefile

12W3
- documentation
- resolve todo in `stan_block_builder`, `stan_model`
-- prevent mixing `statement.lhs_variable vs .lhs_expr`
-- named array (prevent confusion)
```
  # TODO @Dashadower dims[0] VS using Q, R
elif len(dims) == 1:
  self.stan_data[key] = StanDataEntry(key, f"vector[{dims[0]}]")
elif len(dims) == 2:
  self.stan_data[key] = StanDataEntry(key, f"array[{dims[0]}] vector[{dims[1]}]")
```

- auto-coordinate generation, generate (target, _obs) pairs
12W4
- speed

**Reply by Dashadower** [reply]
- loglik plot
- calibrator refactor, but restricted to the scope of implementing loglik
- builder todo

**Reply by Dashadower** [reply]
pdocs support for stanify

**Reply by hyunjimoon** [reply]
12W4
1. difference between two pr? might be better for shinyoung to take the lead on refactoring; how long?
2. loglik plots? defining degeneracy diagnostics for dynamic model 
- multimodal loglik compare: https://github.com/Data4DM/BayesSD/discussions/134

**Reply by hyunjimoon** [reply]
with @Dashadower next week
- sbc rank plot for data-loglik (M by S)
- sbc rank plot with `lp__` and divergence mapping (hypothesis: the two have high negative correlation, meaning low lp__ would have create high divergence from bad geometry

**Reply by hyunjimoon** [reply]
Shinyoung, Angie 1W1
## did
1. plot with xarray dataset (on time coordinate)
2. arviz assign within group and sum between prior_predictive and posterior_predictive

## food for thoughts
### 1. variation ratio to each layer as sbc test quantity
With user-defined heterogeneity decomposition of total heterogeneity (variation) to individual (unit) level and population level e.g. 3:1 =`log_pop`: `log_unit`, can we do sbc with test quantity `.25log_pop-log_unit`?

<img width="897" alt="image" src="https://user-images.githubusercontent.com/30194633/211178923-5a1727b6-950a-4b13-9b46-347c2651a027.png">

### 2. could we relate loglike_unit and loglik_pop with contraction, containment, compromise?
![image](https://user-images.githubusercontent.com/30194633/211179148-4af351c3-be9d-408d-a5a2-d7832f3acedf.png)
i) contraction: unit_layer (likelihood) is more informative, but consistent with population_layer (prior)
loglik_unit << loglik_pop

ii) containment: population_layer (prior) is more informative, but consistent with unit_layer (likelihood) 
loglik_unit >> loglik_pop

iii) compromise: tension btw information encoded in population and unit
loglik_unit ~ loglik_pop


### 3. example for heterogeneity contribution to multilayer (one layer added to unit-population above)
: manufacturer (hyper-population) + engine type (population) + engine (unit)
For forklift data we had from paper below,
<img width="576" alt="image" src="https://user-images.githubusercontent.com/30194633/211178831-a127681d-b3bd-4d60-a0d5-36926e359efe.png">

[Kim22_FailureFuncLogisticsBayesAgg.pdf](https://github.com/Data4DM/BayesSD/files/10367626/Kim22_FailureFuncLogisticsBayesAgg.pdf)

**Reply by hyunjimoon** [reply]
@Dashadower could we try using adjoint method? https://discourse.mc-stan.org/t/slow-hierarchical-ode-model/30001/2

**Reply by hyunjimoon** [reply]
1W2 Shinyoung, Angie agenda
1. merge shinyoung's branch with pdoc branch in pr
2. is it possible to translate vensim subscript?

**Comment by hyunjimoon** [comment]
Possible ideas to develop:
- https://github.com/Data4DM/stanify/issues/6

**Comment by hyunjimoon** [comment]
2W4
- data (exo_demand) + tests
- loglik-based plot, not one theta as in https://gist.github.com/Dashadower/face0ff696d83a13d0ea1624b70c7917
- - meaning https://github.com/Data4DM/stanify/blob/68cde36549923ff29764171ab1dfc955bfb1bae9/stanify/builders/utilities.py#L255?

**Reply by Dashadower** [reply]
https://github.com/Data4DM/stanify/pull/45

**Reply by Dashadower** [reply]
> [Data4DM/stanify#45](https://github.com/Data4DM/stanify/pull/45)

Especially supporting subscripted data/lookup
